app/Libraries/RouteScopesHelper.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace App\Libraries;
+
+use Route;
+
+class RouteScopesHelper
+{
+    public $routes;
+
+    // fills in any missing require-scopes:
+    public function fillMissingMiddlewares()
+    {
+        // pass by reference so middlewares value can be reassigned.
+        foreach ($this->routes as &$route) {
+            $scopes = $route['scopes'];
+            $middlewares = $route['middlewares'];
+
+            $scopesString = presence(implode(',', $scopes));
+            if ($scopesString === null) {
+                continue;
+            }
+
+            // add missing middleware if necessary; exact order might be wrong.
+            $newString = "require-scopes:{$scopesString}";
+            $index = array_search('require-scopes', $middlewares, true);
+            if ($index === false) {
+                $middlewares[] = 'require-scopes';
+            }
+
+            $exists = false;
+            foreach ($middlewares as &$middleware) {
+                // replace existing value if it exists
+                if (starts_with($middleware, 'require-scopes:')) {
+                    $middleware = $newString;
+                    $exists = true;
+                }
+            }
+
+            // otherwise insert new value after require-scopes if possible.
+            if (!$exists) {
+                array_splice($middlewares, $index ? $index + 1 : count($middlewares), 0, $newString);
+            }
+
+            $route['middlewares'] = $middlewares;
+        }
+    }
+
+    public function fromCsv(string $filename)
+    {
+        $csv = array_map('str_getcsv', file($filename));
```

Comment:
does `fgetcsv` not work?

Reply:
this doesn't require dealing with the file handle